{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4e801919-2447-4a77-88cf-1aea09773a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "from Bio import SeqIO\n",
    "from collections import Counter\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "47d890f3-f396-4815-aae4-829443dfc2fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of amino acids and all possible dipeptides\n",
    "amino_acids = \"ARNDCEQGHILKMFPSTWYV\"\n",
    "dipeptides = [a1 + a2 for a1 in amino_acids for a2 in amino_acids]\n",
    "\n",
    "def calculate_amino_acid_percentage(sequence):\n",
    "    \"\"\"Calculate the percentage of each amino acid in the given sequence.\"\"\"\n",
    "    total_length = len(sequence)\n",
    "    amino_acid_counts = Counter(sequence)\n",
    "    return {aa: (amino_acid_counts[aa] / total_length) * 100 if total_length > 0 else 0 for aa in amino_acids}\n",
    "\n",
    "'''\n",
    "def calculate_dipeptide_percentage(sequence):\n",
    "    \"\"\"Calculate the percentage of each dipeptide in the given sequence.\"\"\"\n",
    "    dipeptide_list = [sequence[i:i+2] for i in range(len(sequence) - 1)]\n",
    "    dipeptide_counts = Counter(dipeptide_list)\n",
    "    total_dipeptides = len(dipeptide_list)\n",
    "    return {dp: (dipeptide_counts[dp] / total_dipeptides) * 100 if total_dipeptides > 0 else 0 for dp in dipeptides}\n",
    "'''\n",
    "\n",
    "def calculate_dipeptide_percentage(sequence, g):\n",
    "    \"\"\"Calculate the percentage of each dipeptide in the given sequence.\"\"\"\n",
    "    dipeptide_list = [sequence[i]+sequence[i+g] for i in range(len(sequence) - g)]\n",
    "    dipeptide_counts = Counter(dipeptide_list)\n",
    "    total_dipeptides = len(dipeptide_list)\n",
    "    return {dp: (dipeptide_counts[dp] / total_dipeptides) * 100 if total_dipeptides > 0 else 0 for dp in dipeptides}\n",
    "\n",
    "def aliphatic_index(sequence):\n",
    "    analysed_seq = ProteinAnalysis(sequence)\n",
    "    aa_counts = analysed_seq.count_amino_acids()\n",
    "    \n",
    "    A = aa_counts['A']  # Alanine\n",
    "    V = aa_counts['V']  # Valine\n",
    "    I = aa_counts['I']  # Isoleucine\n",
    "    L = aa_counts['L']  # Leucine\n",
    "    seq_length = sum(aa_counts.values())  # Total amino acids\n",
    "    \n",
    "    # Aliphatic index formula\n",
    "    AI = ((A + 2.9 * V + 3.9 * (I + L)) / seq_length) * 100\n",
    "    return AI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dc037799-d04b-41a1-bccc-a8509659ba55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6-CTD\n",
    "group1 = {\n",
    "    'hydrophobicity_PRAM900101': 'RKEDQN',\n",
    "    'hydrophobicity_ARGP820101': 'QSTNGDE',\n",
    "    'hydrophobicity_ZIMJ680101': 'QNGSWTDERA',\n",
    "    'hydrophobicity_PONP930101': 'KPDESNQT',\n",
    "    'hydrophobicity_CASG920101': 'KDEQPSRNTG',\n",
    "    'hydrophobicity_ENGD860101': 'RDKENQHYP',\n",
    "    'hydrophobicity_FASG890101': 'KERSQD',\n",
    "    'normwaalsvolume': 'GASTPDC',\n",
    "    'polarity':        'LIFWCMVY',\n",
    "    'polarizability':  'GASDT',\n",
    "    'charge':          'KR',\n",
    "    'secondarystruct': 'EALMQKRH',\n",
    "    'solventaccess':   'ALFCGIVW'\n",
    "}\n",
    "group2 = {\n",
    "    'hydrophobicity_PRAM900101': 'GASTPHY',\n",
    "    'hydrophobicity_ARGP820101': 'RAHCKMV',\n",
    "    'hydrophobicity_ZIMJ680101': 'HMCKV',\n",
    "    'hydrophobicity_PONP930101': 'GRHA',\n",
    "    'hydrophobicity_CASG920101': 'AHYMLV',\n",
    "    'hydrophobicity_ENGD860101': 'SGTAW',\n",
    "    'hydrophobicity_FASG890101': 'NTPG',\n",
    "    'normwaalsvolume': 'NVEQIL',\n",
    "    'polarity':        'PATGS',\n",
    "    'polarizability':  'CPNVEQIL',\n",
    "    'charge':          'ANCQGHILMFPSTWYV',\n",
    "    'secondarystruct': 'VIYCWFT',\n",
    "    'solventaccess':   'RKQEND'\n",
    "}\n",
    "group3 = {\n",
    "    'hydrophobicity_PRAM900101': 'CLVIMFW',\n",
    "    'hydrophobicity_ARGP820101': 'LYPFIW',\n",
    "    'hydrophobicity_ZIMJ680101': 'LPFYI',\n",
    "    'hydrophobicity_PONP930101': 'YMFWLCVI',\n",
    "    'hydrophobicity_CASG920101': 'FIWC',\n",
    "    'hydrophobicity_ENGD860101': 'CVLIMF',\n",
    "    'hydrophobicity_FASG890101': 'AYHWVMFLIC',\n",
    "    'normwaalsvolume': 'MHKFRYW',\n",
    "    'polarity':        'HQRKNED',\n",
    "    'polarizability':  'KMHFRYW',\n",
    "    'charge':          'DE',\n",
    "    'secondarystruct': 'GNPSD',\n",
    "    'solventaccess':   'MSPTHY'\n",
    "}\n",
    "groups = [group1, group2, group3]\n",
    "ctd_property = (\n",
    "'hydrophobicity_PRAM900101', 'hydrophobicity_ARGP820101', 'hydrophobicity_ZIMJ680101', 'hydrophobicity_PONP930101',\n",
    "'hydrophobicity_CASG920101', 'hydrophobicity_ENGD860101', 'hydrophobicity_FASG890101', 'normwaalsvolume',\n",
    "'polarity', 'polarizability', 'charge', 'secondarystruct', 'solventaccess')\n",
    "def ctd_composition(sequence):\n",
    "    c = Counter(sequence)\n",
    "    encodings = []\n",
    "    for p in ctd_property:\n",
    "        c1 = sum([c[k] for k in group1[p]])\n",
    "        c2 = sum([c[k] for k in group2[p]])\n",
    "        encodings += [c1, c2]\n",
    "    return encodings\n",
    "    '''\n",
    "    def Count(seq1, seq2):\n",
    "        s = 0\n",
    "        for aa in seq1:\n",
    "            s += seq2.count(aa)\n",
    "        return s\n",
    "    encodings = []\n",
    "    for i in sequence:\n",
    "        for p in ctd_property:\n",
    "            c1 = Count(group1[p], sequence) / len(sequence)\n",
    "            c2 = Count(group2[p], sequence) / len(sequence)\n",
    "            c3 = 1 - c1 - c2\n",
    "            encodings = encodings + [c1, c2, c3]\n",
    "    return encodings\n",
    "    '''\n",
    "transitions = {}\n",
    "total_transitions = {}\n",
    "for p in ctd_property:\n",
    "    transitions[p] = {'group1-group2': 0, 'group1-group3': 0, 'group2-group1': 0, 'group2-group3': 0, 'group3-group1': 0, 'group3-group2': 0}\n",
    "    total_transitions[p] = 0\n",
    "def ctd_transitions(sequence):\n",
    "    for i in range(len(sequence)-1):\n",
    "        a1, a2 = sequence[i], sequence[i+1]\n",
    "        for p in ctd_property:\n",
    "            if a1 in group1[p] and a2 in group2[p]: transitions[p]['group1-group2'] += 1\n",
    "            if a1 in group1[p] and a2 in group3[p]: transitions[p]['group1-group3'] += 1\n",
    "            if a1 in group2[p] and a2 in group1[p]: transitions[p]['group2-group1'] += 1\n",
    "            if a1 in group2[p] and a2 in group3[p]: transitions[p]['group2-group3'] += 1\n",
    "            if a1 in group3[p] and a2 in group1[p]: transitions[p]['group3-group1'] += 1\n",
    "            if a1 in group3[p] and a2 in group2[p]: transitions[p]['group3-group2'] += 1\n",
    "    for p in ctd_property:\n",
    "        total_transitions[p] = sum(transitions[p].values())\n",
    "        #for k in transitions[p]:\n",
    "        #    total_transitions[p] += transitions[p][k]\n",
    "    #total_transitions = sum(transition.values())\n",
    "    #if total_transitions == 0: return transitions\n",
    "    return {p+'.'+k: (v / total_transitions[p])*100 if total_transitions[p] != 0 else 0 for p in ctd_property for k, v in transitions[p].items()}\n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1b282419-1c34-48d4-99c3-8a398a088b94",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_protein_features(sequence):\n",
    "    \"\"\"Calculate protein features using Biopython's ProteinAnalysis.\"\"\"\n",
    "    analyzed_seq = ProteinAnalysis(str(sequence))\n",
    "    AI = aliphatic_index(sequence)\n",
    "\n",
    "    return {\n",
    "        \"Length\": len(sequence),\n",
    "        #\"Molecular Weight\": analyzed_seq.molecular_weight(),\n",
    "        #\"Aromaticity\": analyzed_seq.aromaticity(),\n",
    "        #\"GRAVY\": analyzed_seq.gravy(),\n",
    "        #\"Isoelectric Point\": analyzed_seq.isoelectric_point(),\n",
    "        #\"Helix Fraction\": analyzed_seq.secondary_structure_fraction()[0],\n",
    "        #\"Turn Fraction\": analyzed_seq.secondary_structure_fraction()[1],\n",
    "        #\"Sheet Fraction\": analyzed_seq.secondary_structure_fraction()[2],\n",
    "        #\"hydrophobicity\": sum(analyzed_seq.protein_scale(window=5, param_dict=KyteDoolittle1)) / len(sequence),\n",
    "        #\"instability_index\": analyzed_seq.instability_index(),\n",
    "        #\"extenictioin_coeff_X\": analyzed_seq.molar_extinction_coefficient()[0],\n",
    "        #\"extenictioin_coeff_Y\": analyzed_seq.molar_extinction_coefficient()[1],\n",
    "        ##\"aa_composition\": aa_composition = analysed_seq.count_amino_acids(),\n",
    "        #\"aliphatic_index\": AI,\n",
    "    }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7af2cd5c-186e-4f50-a8b4-bbee23e7c953",
   "metadata": {},
   "outputs": [],
   "source": [
    "maxGap = 3\n",
    "def process_proteomes(proteome_files, proteome_label, output_csv):\n",
    "    \"\"\"Process sequences from each proteome file and append features to CSV.\"\"\"\n",
    "    with open(output_csv, \"a\", newline=\"\") as csvfile:\n",
    "        csvwriter = csv.writer(csvfile)\n",
    "\n",
    "        for file_path in proteome_files:\n",
    "            print(file_path)\n",
    "            for seq_record in SeqIO.parse(file_path, \"fasta\"):\n",
    "                sequence = str(seq_record.seq)\n",
    "\n",
    "                # **Skip sequences that contain 'X'**\n",
    "                #if \"X\" in sequence:\n",
    "                #    continue  # Skip this sequence entirely\n",
    "                #if \"U\" in sequence: \n",
    "                    #u_list.append(seq_record.seq)\n",
    "                #    continue\n",
    "                #if \"n\" in sequence: \n",
    "                    #n_list.append((seq_record.id, file_path))\n",
    "                #    continue\n",
    "                # Process valid sequences\n",
    "                protein_features = calculate_protein_features(sequence)\n",
    "                #aa_percentages = calculate_amino_acid_percentage(sequence)\n",
    "                '''\n",
    "                dipeptide_percentages = {}\n",
    "                for g in range(1, maxGap):\n",
    "                    dipeptide_percentages[g] = calculate_dipeptide_percentage(sequence, g)\n",
    "                '''\n",
    "                ctd_composition_percentages = ctd_composition(sequence)\n",
    "                ctd_transitions_percentages = ctd_transitions(sequence)\n",
    "                \n",
    "                # Combine all features\n",
    "                row = [\n",
    "                    seq_record.id,\n",
    "                    sequence,\n",
    "                    proteome_label,\n",
    "                    protein_features[\"Length\"],\n",
    "                    #protein_features[\"Molecular Weight\"],\n",
    "                    #protein_features[\"Aromaticity\"],\n",
    "                    #protein_features[\"GRAVY\"],\n",
    "                    #protein_features[\"Isoelectric Point\"],\n",
    "                    #protein_features[\"Helix Fraction\"],\n",
    "                    #protein_features[\"Turn Fraction\"],\n",
    "                    #protein_features[\"Sheet Fraction\"],\n",
    "                    #protein_features[\"hydrophobicity\"],\n",
    "                    #protein_features[\"instability_index\"],\n",
    "                    #protein_features[\"extenictioin_coeff_X\"],\n",
    "                    #protein_features[\"extenictioin_coeff_Y\"],\n",
    "                    #protein_features[\"aliphatic_index\"],\n",
    "                ]\n",
    "                #row += [aa_percentages[aa] for aa in amino_acids]\n",
    "                row += ctd_composition_percentages\n",
    "                row += list(ctd_transitions_percentages.values())\n",
    "                '''\n",
    "                for g in range(1, maxGap):\n",
    "                    row + [dipeptide_percentages[g][dp] for dp in dipeptides]\n",
    "                '''\n",
    "                csvwriter.writerow(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1d81716a-fc43-4360-8e92-063b5a8be051",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "108\n",
      "Processing thermophilic proteomes label 0\n",
      "A_aeolicus.txt\n",
      "C_abyssi.txt\n",
      "D_thermophilum.txt\n",
      "P_methylaliphatogenes.txt\n",
      "R_marinus.txt\n",
      "S_thermophila.txt\n",
      "T_ islandicus.txt\n",
      "T_aquaticus.txt\n",
      "T_indicus.txt\n",
      "T_marianensis.txt\n",
      "T_maritima.txt\n",
      "T_velox.txt\n",
      "Processing psychrophilic proteomes label 1\n",
      "A_sublithincola.txt\n",
      "C_gasigenes.txt\n",
      "D_frigens.txt\n",
      "G_pallidula.txt\n",
      "G_superstes.txt\n",
      "H_roseosalivarius.txt\n",
      "M_psychrophila.txt\n",
      "P_agardhii.txt\n",
      "P_atlanticus.txt\n",
      "T_saanensis.txt\n",
      "Processing mesophilic proteomes (label = 2)\n",
      "brachy_murdo.txt\n",
      "c_falvigena.txt\n",
      "deino_mari.txt\n",
      "ecoli.txt\n",
      "emp_brevis.txt\n",
      "flecto_major.txt\n",
      "hol_foetida.txt\n",
      "keton_race.txt\n",
      "lepto_shahi.txt\n",
      "mycoplasma.txt\n",
      "pro_sphenisci_proteom.txt\n",
      "Amino acid, dipeptide, and structural features saved to filtered_proteome_data5.csv.\n"
     ]
    }
   ],
   "source": [
    "# File lists for both types of proteomes\n",
    "thermophilic_files = [\n",
    "    \"A_aeolicus.txt\", \"C_abyssi.txt\", \"D_thermophilum.txt\", \n",
    "    \"P_methylaliphatogenes.txt\", \"R_marinus.txt\", \"S_thermophila.txt\",\n",
    "    \"T_ islandicus.txt\", \"T_aquaticus.txt\", \"T_indicus.txt\",\n",
    "    \"T_marianensis.txt\", \"T_maritima.txt\", \"T_velox.txt\"\n",
    "]\n",
    "\n",
    "psychrophilic_files = [\n",
    "    \"A_sublithincola.txt\", \"C_gasigenes.txt\", \"D_frigens.txt\",\n",
    "    \"G_pallidula.txt\", \"G_superstes.txt\", \"H_roseosalivarius.txt\",\n",
    "    \"M_psychrophila.txt\", \"P_agardhii.txt\", \"P_atlanticus.txt\", \"T_saanensis.txt\"\n",
    "]\n",
    "\n",
    "mesophilic_files = [\n",
    "    \"brachy_murdo.txt\", \"c_falvigena.txt\", \"deino_mari.txt\", \"ecoli.txt\", \n",
    "    \"emp_brevis.txt\", \"flecto_major.txt\", \"hol_foetida.txt\", \"keton_race.txt\",\n",
    "    \"lepto_shahi.txt\", \"mycoplasma.txt\", \"pro_sphenisci_proteom.txt\"\n",
    "]\n",
    "\n",
    "# Output CSV file\n",
    "output_csv = \"filtered_proteome_data5.csv\"\n",
    "\n",
    "# Write the header row to the combined CSV\n",
    "with open(output_csv, \"w\", newline=\"\") as csvfile:\n",
    "    csvwriter = csv.writer(csvfile)\n",
    "    header = [\n",
    "        \"Sequence ID\", \"Seq\", \"Proteome Type (0=Hot, 1=Cold, 2=Meso)\", \"Length\", \n",
    "        #\"Molecular Weight\", \n",
    "        #\"Aromaticity\", \"GRAVY\", \"Isoelectric Point\", \"Helix Fraction\", \"Turn Fraction\", \"Sheet Fraction\",\n",
    "        #\"hydrophobicity\",\"instability_index\", \"extenictioin_coeff_X\", \"extenictioin_coeff_Y\", \"aliphatic_index\",\n",
    "    ]\n",
    "    #header += list(amino_acids) #+ dipeptides\n",
    "    ## ctd_composition header\n",
    "    \n",
    "    for p in ctd_property:\n",
    "        for k in range(1, 3):\n",
    "            header.append(p + '.G' + str(k))\n",
    "    \n",
    "    # ctd_transition header\n",
    "    for p in ctd_property:\n",
    "        for k in transitions[p]:\n",
    "            header.append(p+'.'+k)\n",
    "    '''\n",
    "    for p in range(maxGap):\n",
    "        for k in dipeptides:\n",
    "            header.append(k + str(p))\n",
    "    '''\n",
    "    print(len(header))\n",
    "    csvwriter.writerow(header)\n",
    "\n",
    "print('Processing thermophilic proteomes label 0')\n",
    "# Process thermophilic proteomes (label = 0)\n",
    "process_proteomes(thermophilic_files, 0, output_csv)\n",
    "\n",
    "print('Processing psychrophilic proteomes label 1')\n",
    "# Process psychrophilic proteomes (label = 1)\n",
    "process_proteomes(psychrophilic_files, 1, output_csv)\n",
    "\n",
    "# Process mesophilic proteomes (label = 2)\n",
    "print('Processing mesophilic proteomes (label = 2)')\n",
    "process_proteomes(mesophilic_files, 2, output_csv)\n",
    "\n",
    "print(f\"Amino acid, dipeptide, and structural features saved to {output_csv}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a69b2c1f-64c4-45c2-8614-2b8c94b65445",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12de74e1-7c2c-49fd-8d2a-2b4cf11468f5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
